collector/lib/CollectorConfig.cpp

```diff
@@ -447,6 +451,73 @@ void CollectorConfig::HandleConfig(const std::filesystem::path& filePath) {
   }
 }
 
+void WaitForFileToExist(const std::filesystem::path& filePath) {
+  while (!std::filesystem::exists(filePath)) {
+    sleep(5);
+  }
+}
+
+void CollectorConfig::WatchConfigFile(const std::filesystem::path& filePath) {
+  CLOG(INFO) << "In WatchConfigFile";
+  int fd = inotify_init();
+  if (fd < 0) {
+    CLOG(FATAL) << "inotify_init failed";
+    return;
+  }
+
+  WaitForFileToExist(filePath);
+  HandleConfig(filePath);
+  int wd = inotify_add_watch(fd, filePath.c_str(), IN_MODIFY | IN_MOVE_SELF | IN_DELETE_SELF);
+  if (wd < 0) {
+    CLOG(FATAL) << "Failed to add inotify watch for " << filePath;
+    close(fd);
+    return;
+  }
+
+  char buffer[1024];
+  while (true) {
+    CLOG(INFO) << "Waiting for file event";
+    int length = read(fd, buffer, sizeof(buffer));
+    if (length < 0) {
+      CLOG(FATAL) << "Unable to read event for " << filePath;
+    }
+
+    for (int i = 0; i < length; i += sizeof(struct inotify_event)) {
+      struct inotify_event* event = (struct inotify_event*)&buffer[i];
+      CLOG(INFO) << "Got event event->mask= " << event->mask;
+      CLOG(INFO) << "event->mask & IN_MODIFY= " << (event->mask & IN_MODIFY);
+      if (event->mask & IN_MODIFY) {
+        CLOG(INFO) << "File was replaced";
+        HandleConfig(filePath);
+      } else if ((event->mask & IN_MOVE_SELF) || (event->mask & IN_DELETE_SELF) || (event->mask & IN_IGNORED)) {
+        CLOG(INFO) << "File was removed";
+        WaitForFileToExist(filePath);
+        wd = inotify_add_watch(fd, filePath.c_str(), IN_MODIFY | IN_MOVE_SELF | IN_DELETE_SELF);
+        HandleConfig(filePath);
+        if (wd < 0) {
+          CLOG(FATAL) << "Failed to add inotify watch for " << filePath;
```

Comment:
Similar to my previous comment, I wouldn't kill collector completely in this situation.